Error extensions improvements

glasser opened the pull request:
This PR does a few things:

(a) Moves `error.extensions.exception.stacktrace` to
    `error.extensions.stacktrace`.

(b) Removes the rest of `error.extensions.exception`. This contained
    enumerable properties of the original error, which could lead to
    surprising information leakage.

(c) Documents that change and provides a `formatError` hook
    in the migration guide to maintain AS3 behavior.

(d) Provide an `unwrapResolverError` function in `@apollo/server/errors`
    to help get back the original error in the resolver error case, which is
    a helpful part of the documented `formatError` recommendations
    (since the old error formatting put stuff from that unwrapped error
    on the `exception` extension).

(e) Gets rid of the `declare module` which made
    `error.extensions.exception.code`/`error.extensions.exception.stacktrace`
    have a non-unknown value. Note that this declaration (added in 3.5.0) was
    actually inaccurate as `code` really goes directly on `extensions` rather than
    on `exception`. We could have instead preserved the declaration
    and adapted it to the new location of `stacktrace` and the correct
    location of `code`.

  - Pro: `declare module` is a bit scary and doesn't always merge well if
    you have more than one of them (eg, if you end up with both AS3 and AS4
    in your TypeScript build: AS3 had a different `declare module` where
    `code` was nested under `exception`).

  - Con: End users who catch our errors can't get "correct" typing for
    `error.extensions.code` or `error.extensions.stacktrace`.

  - Pro: That's only "correct" if you assume all GraphQLErrors use
    extensions compatible with the definition; it might be better to export
    a helper function that takes a `GraphQLError` and returns the
    code/exception if it looks like it has the right shape. And nobody seems
    to have even noticed that the declaration has been wrong for almost a
    year, so how much value is it providing?

(f) Renames the (new in v4) constructor option
    includeStackTracesInErrorResponses to
    includeStacktraceInErrorResponses, to match the extension name.

(g) Removes a test around error handling a particular `yup` style of errors
    that is probably not relevant any more.

This is to some degree part of #6719 because we're concerned about the
effect of the `declare module` on the Gateway transition.

IvanGoncharov reviewed (commented):
Yes, declaration merging is problematic so if you think it will cause an issue removing it is the right call.
One possible alternative would be to do a breaking change and just move `stacktrace` at the top level of `extensions`.

glasser commented:
@IvanGoncharov out of curiosity, why does that help? (It does seem reasonable though, esp since we already moved code out.)

glasser commented:
BTW, we need to add the change of `code`'s location to the migration guide (can be done as part of this PR).

glasser commented:
@IvanGoncharov OK, I've made a few more changes along these lines.

IvanGoncharov commented on packages/server/src/__tests__/errors.test.ts:
```diff
@@ -41,36 +41,32 @@ describe('Errors', () => {
-      expect(error.extensions.code).toEqual(code);
-      // stacktrace should exist under exception
-      expect(error.extensions.exception.stacktrace).toBeDefined();
+      expect(error.extensions?.key).toEqual(key);
```
`extensions` are always present, you shouldn't need optional chaining here.

glasser replied:
@IvanGoncharov It's actually optional on GraphQLFormattedError, which is what this is.

IvanGoncharov commented on packages/server/src/errorNormalize.ts:
```diff
@@ -66,17 +67,18 @@ export function normalizeAndFormatErrors(
         extensions.exception = {
-          ...extensions.exception,
+          ...ensureObject(extensions.exception),
           ...Object.fromEntries(originalErrorEnumerableEntries),
```
It's increasing the scope of this PR, but maybe we can remove the behavior of passing all enumerable properties as `extensions.exception`.
This practice can be considered as information leaking, similar to "did you mean".
Now that we gave `formatError` access to the original issue, users can peek and choose what properties they want to put into `extensions`.
So basically, I propose to just get rid of `extensions.exception`.

IvanGoncharov commented:
> @IvanGoncharov out of curiosity, why does that help? (It does seem reasonable though, esp since we already moved code out.)

My suggestion was based on the idea of completely removing `exception`.
I focused solely on `stacktrace` because I forgot that `exception` also contains top-level properties from the thrown error.
So the main idea is to just remove `exception` completely.

glasser commented:
@IvanGoncharov I liked your suggestion and took it.

I updated the PR description, which is worth re-reading.

One thing I found is that documenting that the second argument to `formatError` is the originally thrown error wasn't quite accurate. That's because in the most common case (errors thrown in resolvers), the originally thrown error will be wrapped in a GraphQLError by the graphql-js `locatedError` function. I tried to adjust the migration docs to say that the argument might be the original error or you might have to look on the `originalError` field but that got really weird to explain.

So instead I added a special case to the invocation of `formatError`. If the error it is formatting is a `GraphQLError` with a `path` and an `originalError`, it passes `error.originalError` as the second argument instead. This provides easier access to the *actually* original error. If you need access to `error.path` (or `error.locations`), it's right there on the first `GraphQLFormattedError` argument.

The one downside I see here is that if you wanted access to `nodes` (or `source`) you're out of luck. That's a bit annoying.

Another option is to change `formatError` to take an options argument instead of two arguments, and then provide three arguments:

```
formatError({
  defaultFormattedError: GraphQLFormattedError,
  thrownError: Error,
  locatedError: GraphQLError | null,
})
```

If `error` is the actual error provided to `normalizeAndFormatErrors`, then:

```
defaultFormattedError = enrichError(error)
thrownError = error instanceof GraphQLError && error.path && error.originalError
          ? error.originalError
          : error
locatedError = error instanceof GraphQLError && error.path ? error : null
```

Maybe this is more complicated than it's worth and we should just explain to users that they need to sometimes (but not always!) look at `error.originalError` in their formatError... but I don't love that either.

IvanGoncharov commented on docs/source/migration.mdx:
```diff
@@ -1062,6 +1066,87 @@ So now you can format errors as such:
+  error: unknown,
+) {
+  const exception: Record<string, unknown> = {
+    ...(typeof error === 'object' ? error : null),
```
better to do `instanceof Error` check here.
We need to protect against strings and arrays (pass you current check).
<img width="210" alt="image" src="https://user-images.githubusercontent.com/8336157/183117076-3f0ac117-fdb8-4556-9d29-ae69192d763b.png">

IvanGoncharov replied:
Also to work with `originalError` it can be
```
...(error instanceof Error ? error.originalError ?? error : null),
```

glasser replied:
I think I'm not super-worried about things like strings and arrays here... if you pass something ultra weird, then your code ends up weird. I just want to make sure TS builds.

IvanGoncharov commented on packages/server/src/errorNormalize.ts:
```diff
@@ -26,16 +17,25 @@ export function normalizeAndFormatErrors(
+        // there is no path, this error didn't come from a resolver.)
+        error instanceof GraphQLError && error.path && error.originalError
+          ? error.originalError
+          : error,
```
passing originalError makes sense even if `path` is absent.
example errors thrown from coerceValue/serialize since they also wrapped.
So I recommend dropping the `path` check here.

IvanGoncharov commented:
@glasser I think the cleanest semantic would be that `formatError` gets the most top-level error, and the first argument is serialized version of it and the second argument is a raw error.
Meaning that if you need an original error, you need to look into the chain of `originalError` yourself.
So your example in migration guide would have this code but using second argument:
https://github.com/apollographql/apollo-server/pull/6759/files#diff-b519aadbfa6ea02ae453a6c785bcf1038f7844c43549e9bcb2eed2944df55b48L61-L73

glasser commented:
@IvanGoncharov going with your suggestion, plus a new `unwrapResolverError` function that can be helpful.

trevor-scheer commented on packages/server/src/errors/index.ts:
```diff
@@ -8,3 +10,19 @@ export enum ApolloServerErrorCode {
+ * (like parse errors) are not unwrapped by this function.
+ */
+export function unwrapResolverError(error: unknown): unknown {
+  if (error instanceof GraphQLError && error.path && error.originalError) {
```
Not sure I understand why we also need the `error.path` check here.

Unrelated: I didn't find anything interesting on this path but maybe you will. Pretty sure the incoming `unknown` makes any attempt at refining the return type pointless. Might only be useful in a context where you have some type info about the incoming `error` argument.
```ts
export function unwrapResolverError<T = unknown>(error: T): T | Error {
  if (error instanceof GraphQLError && error.path && error.originalError) {
    return error.originalError;
  }
  return error;
}
```

glasser replied:
@trevor-scheer Well, that's what makes this "unwrapResolverError" and not just "unwrapError".

Look at, say, SyntaxError — it's a GraphQLError defined in our code that wraps a GraphQLError received from graphql-js, adding `code` as an extension. I don't think it's helpful for a formatError to strip off our SyntaxError, is it?

trevor-scheer replied:
Got it, I was only seeing this as a type-level refinement and didn't see why we needed it to access `originalError`.

glasser commented on packages/integration-testsuite/src/apolloServerTests.ts:
```diff
@@ -843,7 +843,7 @@ export function defineIntegrationTestSuiteApolloServerTests(
           },
           introspection: true,
-          includeStackTracesInErrorResponses: true,
+          includeStacktraceInErrorResponses: true,
```
I think I'm going to delete this test. Reviewing the initial PR that introduced it (#1288) I don't understand what was interesting about this random `yup` package and suspect it is related to this flattening code which we have deleted. The PR description refers to schema-stitching stuff too. Since the error code is so much simpler now I doubt we have to test one random style of error.
